[CORL-2504]: implement virtualized rendering

kabeaty opened the pull request:
## What does this PR do?

These changes integrate `react-virtuoso` into the comments stream and update the comment stream to use virtualized rendering.

As part of that, these changes support the Z key and the alternate oldest view. They also update how media embeds are shown (default size that can then be expanded to be viewed in full).

Also adds a new config option to the admin that controls whether comments are shown in infinite scroll or with one `Load all comments` button after the initial 20 comments.

These changes also update a few test files to use RTL.

## What changes to the GraphQL/Database Schema does this PR introduce?

This adds `loadAllComments` to the schema, to support the configuration of whether comments should load in infinite scroll or whether a Load all comments button should be shown first.

## Does this PR introduce any new environment variables or feature flags? 

no

## If any indexes were added, were they added to `INDEXES.md`?

n/a

## How do I test this PR?

Check out and interact with the stream. Add new comments, replies, change sort order, etc.

Test out Z key with Z_KEY and SEEN_COMMENTS enabled. Test Z key to root comments, replies, looping around from the end of the stream back to the beginning, with new comments via subscription, with new replies via subscription, with the Show more replies button when a comment has many replies, from a null current comment with traversal focus.

When you scroll comments up and out of view, they should be marked as seen with Z_KEY enabled. The unmark all button / shift + A should mark all comments on the story as seen, instead of just all of those in view.

Test out alternate oldest view and clicking `Add comment` takes you to the post comment form below the `Load all comments` button (or initial comments if fewer than 20). Test new comments via subscription in alternate oldest view.

Test out comments with gifs, Tweets, Youtube media, and embedded images. Test out the new expansion for this embedded media. Test out hiding/showing it.

Test out the new config in the admin around whether to show comments in infinite scroll or with a Load all comments button.
 
## How do we deploy this PR?

This requires that the story tree backend work is already deployed and story trees generated, for the Z key portion of these changes.

alinhle committed:
wip

alinhle committed:
feat: mvp coral outside of iframe

alinhle committed:
wip StreamEmbed

alinhle committed:
wip: bring back more functionality

alinhle committed:
wip: Furthe bring back functionality

alinhle committed:
wip: reduce global polyfills, temporary public path, ie11 support, cors

alinhle committed:
revert removing customCSS functionality

alinhle committed:
wip: integrate shadow dom and css

alinhle committed:
wip: inject additional link tags from async chunks

alinhle committed:
wip: fix changeLocale, integrate client mounted routes into dev, expose manifest/staticConfig data

alinhle committed:
wip: bring back staticConfig, localization, custom css, static uri, prevent unstyled content flashing

alinhle committed:
wip: apply bodyClassName inside shadow dom, improve unstyled loading

alinhle committed:
wip: bugfixes, rm pym, fix tests, use staticconfig

alinhle committed:
wip: Fix ClickOutside, fix dev env, fix shadow root error, improve dev experience, allow custom fonts and disabling typography

alinhle committed:
feat: retry loading /embed/bootstrap

alinhle committed:
chore: deprecate bodyClassName for containerClassName

alinhle committed:
fix: tests

alinhle committed:
feat: intelligently preload assets when nearing embed

alinhle committed:
chore: remove support for ie11, adjust supported browsers

alinhle committed:
feat: allow disable individual polyfills

alinhle committed:
fix: tests

alinhle committed:
chore: stream not as a separte build

alinhle committed:
fix: timestamping

alinhle committed:
fix: manifest with missing integrity

alinhle committed:
chore: use same css extract mechanism for prod and dev

alinhle committed:
fix: remove polyfill config, type CoralStream, fix snapshots

alinhle committed:
fix webpack public path, static uri, <Popup />, cors issues, rename #coral-app-container to #coral

alinhle committed:
feat: MatchMedia now matches width of Shadow DOM, support CSS breakpoint classes

alinhle committed:
feat: put MobileToolbar in its own derived Shadow DOM

alinhle committed:
feat: add deprecation message for bodyClassName

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

alinhle committed:
fix: call onload on webpack assets

alinhle committed:
chore: upgrade webpack-assets-manifest

alinhle committed:
chore+fix: rm accidentally commited files, enable events logging per default during dev

alinhle committed:
chore: adapt browser versions

alinhle committed:
chore: refactor constants AUTORENDER_ROOT_MARGIN and PRELOAD_ROOT_MARGIN

alinhle committed:
fix: correct breakpoints for derived shadow doms that include modals

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

alinhle committed:
fix: pass static config

alinhle committed:
fix: populate static config and fix keyboard shortcuts

alinhle committed:
fix: set rootURL

alinhle committed:
fix: immediately update amp height

alinhle committed:
fix: test

alinhle committed:
fix: add comment

alinhle committed:
fix: apply review

alinhle committed:
fix: tests from review

alinhle committed:
fix: graphql cors

alinhle committed:
feat: RTE in iframe mvp

alinhle committed:
refactor: Introduce EncapsulationContext

alinhle committed:
refactor+feat: refactor iframe+shadowroot into common encapsulation concept + improvements, CoralContainer with breakpoints now ubiquitous

alinhle committed:
chore: improve comment

alinhle committed:
fix: improve getCurrentScriptOrigin

alinhle committed:
fix: support older contentRec ResizeObserverEntry Prop

alinhle committed:
fix: cyclic dep

alinhle committed:
fix: CSSAssets onload should only called once, fix iframe margin

alinhle committed:
fix: set correct port in webpack dev server

alinhle committed:
fix: allow cors port websockets in development

alinhle committed:
feat+chore: improve iframe handling, asset loading, and error display

alinhle committed:
feat: replace pym with frame bundle

alinhle committed:
fix: no need to split frame bundle into chunks

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

alinhle committed:
fix: remove iframe scrolling

alinhle committed:
fix: assign unique jsonpFunction to our webpack bundles

alinhle committed:
chore: removing some remaining pym and iframe references

alinhle committed:
chore: dont crash server so quickly

alinhle committed:
fix: use full url for frame media

alinhle committed:
chore: workaround for half pollyfilled requestIdleCallback functionality...

alinhle committed:
fix: responsive issue

alinhle committed:
fix: styles not applied

alinhle committed:
fix: scrolling

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

kabeaty committed:
initial virtuoso

kabeaty committed:
update without commented out

kabeaty committed:
remove more commented out

kabeaty committed:
update deps

kabeaty committed:
update for tests

kabeaty committed:
remove commented-out

kabeaty committed:
import order

kabeaty committed:
remove loadmore test snapshot

kabeaty committed:
add chunks maxSize

kabeaty committed:
initial unseen comment keyboard shortcuts with next

kabeaty committed:
wait for rootcomment after scroll

kabeaty committed:
update rootCommentElement check

kabeaty committed:
initial scrollseekplaceholder

kabeaty committed:
refactor plus z key working with showallreplies

kabeaty committed:
update scrollseekplaceholder

kabeaty committed:
add comments

kabeaty committed:
add loadallcomments button; support for new comments; fix tests

kabeaty committed:
set max old space size in dockerfile

kabeaty committed:
add comments

kabeaty committed:
loadAllComments update plus fix test

kabeaty committed:
update loadAllComments settings

kabeaty committed:
add some comments

kabeaty committed:
update allChildComments input

kabeaty committed:
up scroll velocity for placeholders to 300

nick-funk commented on Dockerfile:
```diff
@@ -1,5 +1,7 @@
 FROM node:14-alpine
 
+ENV NODE_OPTIONS=--max-old-space-size=4096
```
We seem to have this set to 8 GB on CircleCI, do we want to make that update here as well, just to be safe?

`.circleci/config.yml` (Ln 8)
```
NODE_OPTIONS: "--max-old-space-size=8192"
```

kabeaty committed:
Merge branch 'develop' into spike/virtuoso

kabeaty committed:
initial updates to alternate oldest post comment

kabeaty committed:
more updates to alternate oldest first post comment

kabeaty committed:
update snapshot

kabeaty committed:
update alternate oldest view adding new comments

kabeaty committed:
update tests

kabeaty committed:
cleanup; comments; import order

kabeaty committed:
fix post new alternate oldest comments and show load all

kabeaty committed:
update naming

kabeaty committed:
increase scroll seek velocity

kabeaty committed:
update expanded media settings show/hide

kabeaty committed:
increase scroll velocity and viewport

kabeaty committed:
increase scroll velocity a bit more

kabeaty committed:
fix scroll to post comment for alternate oldest

kabeaty committed:
update postcomment scroll

kabeaty committed:
fixed height that expands for embedded tweets/media

kabeaty committed:
fix C key circles around from end of stream

kabeaty committed:
make expand button stream color

kabeaty committed:
migrate general admin config tests to RTL

kabeaty committed:
new comments show from subscription for oldest first

kabeaty committed:
update liveTopLevelComments tests

kabeaty committed:
refactor Frame and add localization

kabeaty committed:
more refactors

kabeaty committed:
comment and style updates

nick-funk committed:
Merge remote-tracking branch 'origin/develop' into spike/virtuoso

# Conflicts:
#	src/core/client/stream/test/comments/create.ts

nick-funk committed:
Show expand button on Tweets in moderation cards

nick-funk committed:
Fix import order in test

nick-funk committed:
Initial implementation of story tree generation

nick-funk committed:
Implement story tree updating and generation

nick-funk committed:
Remove debugging console.log's

nick-funk committed:
Remove double await from auto-correct

nick-funk committed:
Create nextUnseenComment query

nick-funk committed:
Enforce Z_KEY for nextUnseenComment endpoint

nick-funk committed:
Limit generating story tree's to administrators

nick-funk committed:
Remove retrieval of story tree when unnecessary

This saves on bandwidth and serialization of the document since the
tree can be quite large.

nick-funk committed:
Make nextUnseenComment payload id's optional

They can come back as null, so this makes sense.

nick-funk committed:
Fix typo in story find query

nick-funk committed:
Create mutation to regenerate all story trees

nick-funk committed:
Properly resolve the rootCommentID for root comments

nick-funk committed:
Fix bad copied sort order cursor traversals

nick-funk committed:
Remove marking seen on pagination/loading replies

We will only mark seen using the `markCommentsAsSeen` mutation

nick-funk committed:
Make `nextUnseenComment` traverse relative to current comment id

nick-funk committed:
Add a bit more documentation to the next unseen search

kabeaty committed:
next unseen comment fetch

kabeaty committed:
add index to payload; check comment authorID against userID

kabeaty committed:
update to handle null currentCommentID

nick-funk committed:
Filter out non-visible comments from the stack during next unseen

We want to recreate a linear stack that matches the stream as closely
as possible.

nick-funk committed:
Prune the story tree for visibility before searching for next unseen

nick-funk committed:
Properly flatten the story tree and append root indices

nick-funk committed:
Flatten the story according to the stream sort order

nick-funk committed:
Account for viewNew comments stream side when calculating root indices

Also notifies the user if we looped around in the stream when finding
the next unseen and they need to load in the `View [n] new comments`

nick-funk committed:
Store the `viewNewCount` in local so keyboard shortcuts can use it

nick-funk committed:
Set viewNewCount in local to null when we switch stream tabs

nick-funk committed:
Update findNextUnseenVisibleCommentID traversal documentation

nick-funk committed:
Add documentation to new story tree mutation/query types

nick-funk committed:
Fix off by one error in comment entered subscription `viewNewCount`

nick-funk committed:
Fix ordering of story tree root/replies

- Story tree is stored in newest first (CREATED_AT_DESC)
  - This is the same for root level and replies
- Flatten tree will properly flatten root level comments
  according to the sort order given.
  - If you order by CREATED_AT_DESC, it will flatten from
    0 to n (forwards) in the root level tree comments.
  - If you order by CREATED_AT_ASC, it will flatten from
    n to 0 (backwards) in the root level tree comments.
- Flatten tree will properly flatten replies with oldest
first as the sort order does not affect replies, they are
always oldest first (CREATED_AT_ASC) order.

nick-funk committed:
Fix regenerateStoryTrees so it processes all stories properly

kabeaty committed:
dont mark new comments/replies as seen; fix linting errors

nick-funk committed:
Process the story tree regeneration using bulk operations

nick-funk committed:
Add story tree + next unseen features to feat branch from spike/story-tree

We inter-leaved the front-end and back-end changes in the spike branch
while exploring our options, now we need just the back-end changes since
they will be deployed first.

kabeaty committed:
z key changes

kabeaty committed:
uncomment out dependencies in dep array

kabeaty committed:
update nextUnseenCommentFetch; remove console logs

kabeaty committed:
update for load all replies; load all comments; current virtuoso index

kabeaty committed:
add more comments

kabeaty committed:
add more comments

kabeaty committed:
more comments; fix tests

kabeaty committed:
cleanup and comments

kabeaty committed:
remove intersection context changes

kabeaty committed:
add more comments

kabeaty committed:
refactor

nick-funk committed:
Set nextUnseen cursor same as in spike versions

nick-funk committed:
Update story tree state when comments are created/edited/approved/rejected

kabeaty committed:
refactor viewNewCount

nick-funk committed:
Add missing documentation to schema

nick-funk committed:
Generate story tree when a story is unarchived

nick-funk committed:
Improve documentation around story tree / next unseen

nick-funk committed:
Fix typo in `executeBulkStoryTreeWrites` name

kabeaty committed:
Merge branch 'feat/CORL-2560-story-tree' into spike/virtuoso

kabeaty committed:
update for twitter/youtube embeds

kabeaty committed:
some refactors

kabeaty committed:
clean up from merge; update viewNewRepliesCount

kabeaty committed:
Merge branch 'develop' into spike/virtuoso

kabeaty committed:
implement mark seen on scroll comment up out of view

kabeaty committed:
cleanup and linting fix

kabeaty committed:
mark permalink comments seen if at bottom

kabeaty committed:
refactor; update thresholds and snapshots

kabeaty committed:
update dep arrays

kabeaty committed:
update threshold

kabeaty committed:
Merge branch 'develop' into spike/virtuoso

kabeaty committed:
update expand button width

kabeaty committed:
fetch next unseen when oldest first has new comments too

kabeaty committed:
updateButtonStates should disabled based on whether nextUnseenComment or not

kabeaty committed:
update nextZAction

kabeaty committed:
Merge pull request #3974 from coralproject/feat/CORL-2573-mark-comments-read-scroll-up

[CORL-2573]: Mark comments as read on scroll up out of view and at bottom

kabeaty committed:
make sure z action button enabled if is nextunseenComment

kabeaty committed:
just update buttonStates when nextUnseenComment updates

kabeaty committed:
check if nextunseen loaded in virtuoso before scrolling

kabeaty committed:
update for linting

kabeaty committed:
add missing dep from dep arr

kabeaty committed:
updates to interaction of z key / alternate oldest / load all comments buttons

kabeaty committed:
update displayLoadAllButton logic

kabeaty committed:
remove displayloadall check

nick-funk committed:
Only regenerate story trees for unarchived stories

nick-funk committed:
Merge branch 'feat/CORL-2560-story-tree' into spike/virtuoso

kabeaty committed:
update how mark all as seen works

kabeaty committed:
use eventEmitter to update nextUnseen

kabeaty committed:
only use eventemitter for comment seen and traversal focus

kabeaty committed:
decrease defaultUnexpandedHeight

kabeaty committed:
fix linting issue

kabeaty committed:
update all levels of replies on comment to seen for unmark all

kabeaty committed:
update findNextUnseen effect

kabeaty committed:
cleanup unmark all to not query in root

kabeaty committed:
update youtube width & default height based on mobile

nick-funk commented on src/core/server/models/seenComments/seenComments.ts:
```diff
@@ -84,9 +84,21 @@ export async function markSeenComments(
+    const markAllCommentIDs = (
+      await mongo
+        .comments()
+        .find({ storyID }, { projection: { id: 1 } })
```
This needs to filter by `tenantID` as well. It's a small change, so I made it and pushed it up.

nick-funk committed:
Filter by tenantID as well as storyID when retrieving comments

nick-funk reviewed (approved):
Approving again, nice work!

I updated the `markSeenComments` action query when `markAllAsSeen` is set to also filter by `tenantID`. Was a one line change, didn't feel necessary to comment and make you update it when I was already in there and is such a minor tweak. It's simply a safety check, the query would still have worked, but we want to always also filter by `tenantID` just to be safe.

kabeaty committed:
Merge branch 'spike/virtuoso' of github.com:coralproject/talk into spike/virtuoso

kabeaty committed:
add scrollcommentupoutofview event and emit

kabeaty committed:
update scrollCommentUpOutOfView event

kabeaty committed:
cleanup

kabeaty committed:
update oldest first check

marcushaddon reviewed (commented):
meant to leave this comment on another PR, disregard
